python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -159,3 +179,143 @@ def ensure_freshness_checks(checks: Sequence[AssetChecksDefinition]) -> None:
+) -> Optional[str]:
+    """Get the latest complete partition key for the given cron schedule."""
+    if not partitions_def:
+        return None
+    latest_cron_tick_time = check.not_none(
+        get_latest_completed_cron_tick(
+            freshness_cron=freshness_cron,
+            current_time=current_time,
+            timezone=freshness_cron_timezone,
+        )
+    )
+    time_window = check.not_none(partitions_def.get_prev_partition_window(latest_cron_tick_time))
+    return partitions_def.get_partition_key_range_for_time_window(time_window).start
+
+
+def build_freshness_check(
+    asset_key: AssetKey,
+    freshness_cron: Optional[str],
+    freshness_cron_timezone: str,
+    severity: AssetCheckSeverity,
+    lower_bound_delta: datetime.timedelta,
+    asset_property_enforcement_lambda: Optional[Callable[[AssetsDefinition], bool]],
+) -> AssetChecksDefinition:
+    @asset_check(
+        asset=asset_key,
+        description="Evaluates freshness for targeted asset.",
+        name="freshness_check",
+    )
+    def the_check(context: AssetExecutionContext) -> AssetCheckResult:
+        if asset_property_enforcement_lambda:
+            asset_property_enforcement_lambda(
+                context.job_def.asset_layer.asset_graph.get(asset_key).assets_def
+            )
+        current_time = pendulum.now(
+            tz=freshness_cron_timezone
+        )  # Will be replaced with "effective_timestamp" once it's available.
+
+        # Explicit call to partitions def here will be replaced with AssetSlice reference once it's available.
+        partitions_def = cast(
+            Optional[TimeWindowPartitionsDefinition],
+            context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+        )
+        check.invariant(
+            partitions_def is None or isinstance(partitions_def, TimeWindowPartitionsDefinition),
+            "Expected partitions_def to be time-windowed.",
+        )
+
+        last_update_time_lower_bound = get_last_update_time_lower_bound(
+            freshness_cron=freshness_cron,
+            freshness_cron_timezone=freshness_cron_timezone,
+            current_time=current_time,
+            lower_bound_delta=lower_bound_delta,
+        )
+        expected_partition_key = get_latest_complete_partition_key(
+            freshness_cron=freshness_cron,
+            current_time=current_time,
+            freshness_cron_timezone=freshness_cron_timezone,
+            partitions_def=partitions_def,
+        )
+        latest_record = retrieve_latest_record(
+            instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+        )
+        update_time = get_last_updated_timestamp(latest_record)
+        passed = update_time is not None and update_time >= last_update_time_lower_bound.timestamp()
+        partition_descriptor = (
+            f"Partition {expected_partition_key}" if expected_partition_key else "Asset"
```

Comment:
For non-partitioned assets, will the description include the word "partition"? I think we should avoid this.

Reply:
no, as if the expected partition is none, we just call it "Asset"

Reply:
Ah my bad for missing.

Separate but related: the earlier version has more informative descriptions, which I think are worth preserving here.